smtpcode.c

```diff
@@ -0,0 +1,60 @@
+#include "smtpcode.h"
+
+#include "substdio.h"
+#include "stralloc.h"
+
+#define HUGESMTPTEXT 5000
+
+extern stralloc smtptext;
+extern substdio smtpfrom;
+
+static void get(unsigned char *uc)
+{
+  char *ch = (char *)uc;
+  substdio_get(&smtpfrom,ch,1);
+  if (*ch != '\r')
+    if (smtptext.len < HUGESMTPTEXT)
+     if (!stralloc_append(&smtptext,ch)) temp_nomem();
+}
+
+static int get_digit(unsigned char *uc)
+{
+  get(uc);
+  if (*uc < '0' || *uc > '9')
+    return 1;
+  return 0;
+}
+
+unsigned long smtpcode()
+{
+  unsigned char ch;
+  unsigned long code;
+  int err = 0;
+
+  if (!stralloc_copys(&smtptext,"")) temp_nomem();
+
+  err += get_digit(&ch); code = ch - '0';
+  /* valid SMTP codes are >= 200 */
+  if (ch < '2')
+    err = 1;
+  err += get_digit(&ch); code = code * 10 + (ch - '0');
+  err += get_digit(&ch); code = code * 10 + (ch - '0');
```

Comment:
I can propose an alternative `get_digit(int *n)` that returns 1 (err) in case input is not a digit *or* there was a read error, and then cut the need to add `ch - '0'` every time

Reply:
another alternate is to get 3 bytes and use scan_ulong(). e.g.
scan_ulong("a22", &t), t will be 0
scan_ulong("2z2", &t), t will be 2
scan_ulong("22g", &t), t will be 22

```
void
get(char *ch)
{
  substdio_get(&smtpfrom, ch, 1);
  if (*ch != '\r') {
    if (smtptext.len < HUGESMTPTEXT) {
      if (!stralloc_append(&smtptext, ch))
        temp_nomem();
    }
  }
}

int
get_3digit()
{ 
  char            str[4];
  unsigned long   code;

  substdio_get(&smtpfrom, str, 3);
  str[3] = 0;
  scan_ulong(str, &code);
  if (smtptext.len < HUGESMTPTEXT) {
    if (!stralloc_catb(&smtptext, str, 3))
      temp_nomem();
  }
  return code;
}

unsigned long
smtpcode()
{ 
  unsigned char   ch;
  unsigned long   code;
  int             err = 0;

  if (!stralloc_copys(&smtptext, ""))
    temp_nomem();
  code = get_3digit();
  if (code < 200)
    err = 1;
  for (;;) {
    get((char *) &ch);
    if (ch != ' ' && ch != '-')
      err = 1;
    if (ch != '-')
      break;
    while (ch != '\n')
      get((char *) &ch);
    if (get_3digit() != code)
      err = 1;
  }
  while (ch != '\n')
    get((char *) &ch);
  return err ? 999 : code;
}
```

Reply:
What about just changing the comment to
```c
/* valid SMTP codes are >= 200, this is the first digit of it */
```